### PR TITLE
feat: [IDP-373]: Delegate communication changes

### PR DESCRIPTION
**Pull request by vikyathharekal**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47415)

### CODE REVIEW

File: idp-service/src/main/java/io/harness/idp/proxy/delegate/DelegateProxyRequestForwarder.java [hunk 1]
```diff
@@ -87,11 +88,13 @@ public HttpStepResponse forwardRequestToDelegate(String accountId, String url, L
         ErrorNotifyResponseData errorNotifyResponseData = (ErrorNotifyResponseData) responseData;
         log.error("errorMessage: {}", errorNotifyResponseData.getErrorMessage());
       }
+      log.info("responseData httpResponse before typecast: {}", responseData);
       if (responseData instanceof HttpStepResponse) {
         httpResponse = (HttpStepResponse) responseData;
 
-        log.debug("responseData header: {}", httpResponse.getHeader());
-        log.debug("responseData body: {}", httpResponse.getHttpResponseBody());
+        log.info("responseData header: {}", httpResponse.getHeader());
```
Comment (vikyathharekal) [inline idp-service/src/main/java/io/harness/idp/proxy/delegate/DelegateProxyRequestForwarder.java:95]: I'll revert this back to debug log and just print the required things in the info log later.
Right now this is needed for debugging on pre-qa

File: idp-service/src/main/java/io/harness/idp/app/IdpModule.java [hunk 2]
```diff
@@ -254,7 +254,7 @@ public boolean isNotPrimary() {
     install(new AbstractWaiterModule() {
       @Override
       public WaiterConfiguration waiterConfiguration() {
-        return WaiterConfiguration.builder().persistenceLayer(WaiterConfiguration.PersistenceLayer.MORPHIA).build();
+        return WaiterConfiguration.builder().persistenceLayer(WaiterConfiguration.PersistenceLayer.SPRING).build();
```
Comment (sandeepa-kv-jois) [inline idp-service/src/main/java/io/harness/idp/app/IdpModule.java:257]: Why is this change needed?

Comment (sandeepa-kv-jois) [inline idp-service/src/main/java/io/harness/idp/app/IdpModule.java:257]: Why is this change needed?

**Comment by vikyathharekal** [comment]
trigger uts

**Review by sathish-soundarapandian — APPROVED** [review]

File: idp-service/src/main/java/io/harness/idp/app/IdpModule.java (hunk 2 above)
Comment (vikyathharekal) [reply idp-service/src/main/java/io/harness/idp/app/IdpModule.java:257]: So the delegate task response will be stored in idp-service. The wait engine must know the persistence layer that is used in our service, which is spring data.
Screenshot 2023-05-11 at 11 50 33 AM

**Review by sandeepa-kv-jois — APPROVED** [review]

**Comment by vikyathharekal** [comment]
trigger review

**Comment by vikyathharekal** [comment]
trigger uts

**Comment by vikyathharekal** [comment]
trigger uts

**Review by sandeepa-kv-jois — APPROVED** [review]

**Review by sathish-soundarapandian — APPROVED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger review

**Comment by sathish-soundarapandian** [comment]
trigger ut4
